crates/conjure_core/src/rule.rs

```diff
@@ -0,0 +1,36 @@
+use std::fmt::{self, Display, Formatter};
+
+use thiserror::Error;
+
+use crate::ast::Expression;
+
+#[derive(Debug, Error)]
+pub enum RuleApplicationError {
+    #[error("Rule is not applicable")]
+    RuleNotApplicable,
+}
+
+#[derive(Clone, Debug)]
+pub struct Rule<'a> {
+    pub name: &'a str,
```

Comment:
We can see how things progress, but I can imagine this lifetime might get annoying, when you could just let the Rule have a String (particularly if rules start getting dynamically built names).

Can see how things go, and if it gets annoying later.

Reply:
This is because it used to be just `'static`, but we thought it might be useful for it to not always be static for testing.

(I imagine in most useful cases it will just be fully static though!)

It has to be static because most rules will be allocated at compile time through the macro system.